src/librustc_data_structures/smallvec.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! A vector type that allows storing small values directly on the stack.
+
+use std::iter::FromIterator;
+use std::ptr::drop_in_place;
+use std::ops::Deref;
+use std::fmt;
+
+pub struct SmallVec8<T> {
+    count: usize,
+    values: [ManuallyDrop<T>; 8]
+}
+
+impl<T> fmt::Debug for SmallVec8<T>
+    where T: fmt::Debug {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self[..].fmt(f)
+    }
+}
+
+impl<T> Deref for SmallVec8<T> {
+    type Target = [T];
+    fn deref(&self) -> &Self::Target {
+        let values = &self.values[..self.count];
+        unsafe { &*(values as *const [_] as *const [T]) }
+    }
+}
+
+#[allow(unions_with_drop_fields, dead_code)]
+union ManuallyDrop<T> {
```

Comment:
Since rustc devs are the UB police, is this guaranteed to be convertible to `T`, even without `#[repr(C)]`? (I'm reading this PR with interest for the future of the external arrayvec and smallvec crates of course.)